File: 40_combination_sum_II/combinationSum2.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype List[List[int]]
        """
        candidates = sorted(candidates)
        result = []
        
        def dfs(stack, remain, start):
            if remain == 0 and stack not in result:
                result.append(stack)
            
            for idx, item in enumerate(candidates[start:]):
                if remain < item: return
                #if item in candidates[0: start]: continue
                dfs(stack + [item], remain - item, start + idx + 1)          

        dfs([], target, 0)

        return result

if __name__ == '__main__':
    sol = Solution()
    candidates = [10, 1, 2, 7, 6, 1, 5]
    target = 8
    print sol.combinationSum2(candidates, target)

File: 40_combination_sum_II/combination_Sum2.cpp
#include <iostream>
#include <vector>

using namespace std;


class Solution
{
public:
	vector<vector<int> > combinationSum2(vector<int>& candidates, int target)
	{
		vector<vector<int> > result;

		sort(candidates.begin(), candidates.end());

		vector<int> stack;
		//vector<int>::iterator first = candidates.begin();

		recurse(stack, result, candidates, target, 0);


		for(int i = 0; i < result.size(); i++)
		{
			sort(result[i].begin(), result[i].end());
		}

		sort(result.begin(), result.end());
		result.erase(unique(result.begin(), result.end()), result.end());

		return result;
	}
private:
	void recurse(vector<int>& stack, vector<vector<int> >& result, vector<int>& candidates, int target, int start)
	{
		if(start >= candidates.size() && target > 0) return;

		int previous = -1;
		//int i = start;

		if(target == 0)
		{
			result.push_back(stack);
			return;
		}

		//if(i > candidates.size()) return;

		for(int i = start; i < candidates.size(); ++i)
		{
			if(candidates[i] == previous) continue;

			if(target < candidates[i]) return;

			previous = candidates[i];

			//cout << i << " " << previous << endl;

			stack.push_back(candidates[i]);
			recurse(stack, result, candidates, target - candidates[i], start + 1);
			stack.pop_back();
		}

	}
};



int main(int argc, char const *argv[])
{
	Solution sol;
	vector<vector<int> > result;
	int target = 4;
	int arr[] = {1, 2};
	vector<int> nums(arr, arr + (sizeof(arr) / sizeof(*arr)));
	result = sol.combinationSum2(nums, target);
	//cout << result.size() << endl;


	for(int j = 0; j < result.size(); ++j)
	{
		for(int i = 0; i < result[j].size(); ++i)
		{
			cout << result[j][i] << " ";
		}
		cout << endl;
	}


	//cout << result[0].size() << endl;
	return 0;
}